platform/files.go

```diff
@@ -279,38 +283,79 @@ type RunImageMetadata struct {
 	Mirrors []string `toml:"mirrors" json:"mirrors,omitempty"`
 }
 
-func (rm *RunImageMetadata) BestRunImageMirror(registry string) (string, error) {
+//go:generate mockgen -package testmock -destination ../testmock/access_checker.go github.com/buildpacks/lifecycle/platform AccessChecker
+type AccessChecker interface {
+	CheckReadAccess(repo string, keychain authn.Keychain) (bool, error)
+}
+
+type AccessCheckerImpl struct{}
+
+var _ AccessChecker = &AccessCheckerImpl{}
+
+func (a *AccessCheckerImpl) CheckReadAccess(repo string, keychain authn.Keychain) (bool, error) {
+	img, err := remote.NewImage(repo, keychain)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse image reference")
+	}
+
+	return img.CheckReadAccess(), nil
+}
+
+func (rm *RunImageMetadata) BestRunImageMirror(registry string, accessChecker AccessChecker) (string, error) {
 	if rm.Image == "" {
 		return "", errors.New("missing run-image metadata")
 	}
+
 	runImageMirrors := []string{rm.Image}
 	runImageMirrors = append(runImageMirrors, rm.Mirrors...)
-	runImageRef, err := byRegistry(registry, runImageMirrors)
+
+	keychain, err := auth.DefaultKeychain(runImageMirrors...)
 	if err != nil {
-		return "", errors.Wrap(err, "failed to find run image")
+		return "", errors.Wrap(err, "unable to create keychain")
 	}
-	return runImageRef, nil
-}
 
-func (sm *StackMetadata) BestRunImageMirror(registry string) (string, error) {
-	return sm.RunImage.BestRunImageMirror(registry)
-}
+	runImageRef := byRegistry(registry, runImageMirrors, accessChecker, keychain)
+	if runImageRef != "" {
+		return runImageRef, nil
+	}
+
+	for _, image := range runImageMirrors {
+		ok, err := accessChecker.CheckReadAccess(image, keychain)
+		if err != nil {
+			return "", err
+		}
 
-func byRegistry(reg string, imgs []string) (string, error) {
-	if len(imgs) < 1 {
-		return "", errors.New("no images provided to search")
+		if ok {
```

Comment:
Is there a world where we need to gate this new behavior behind a Platform API check?

Reply:
I don't think so. The only changed behaviour would be that a run image was chosen that was not readable (e.g. some registry, but different repo) and in this case another run image might now work. I would rather see it as improvement/bugfix. The check would make sure that the user receives an error or am I missing something? 

WDYT?

Reply:
I think I'm ok with this. @natalieparellano may feel differently though 😄

Reply:
I agree, this should be safe to back-port to previous Platform APIs